CanonicalABI.md

```diff
@@ -0,0 +1,976 @@
+
+```
+def read_from_memory(direction, type, ptr):
+  # Note that like `lift`, no validation happens here.
+  match interface_type
+    f32 => read_4_bytes(ptr),
+    f64 => read_8_bytes(ptr),
+    s8 => read_1_byte(ptr),
+    u8 => read_1_byte(ptr),
+    s16 => read_2_bytes(ptr),
+    u16 => read_2_bytes(ptr),
+    s32 => read_4_bytes(ptr),
+    u32 => read_4_bytes(ptr),
+    s64 => read_8_bytes(ptr),
+    u64 => read_8_bytes(ptr),
+    char => read_4_bytes(ptr),
+
+    list $ty => {
+      ptr = read_4_bytes(ptr)
+      len = read_4_bytes(ptr + 4)
+      lift(direction, type, [ptr, len])
+    }
+
+    record $fields => {
+      if is_bitflags($fields):
+        if $fields.len() > 32:
+          lift(direction, type, [read_8_bytes(ptr)])
+        else:
+          lift(direction, type, [read_4_bytes(ptr)])
+
+      else:
+        fields = []
+        for field in $fields:
+          ptr = align_up_to(ptr, align(direction, field))
+          fields.push(read_from_memory(direction, field, ptr))
+          ptr += size(direction, field)
+        fields as $language_struct
+    }
+
+    variant $cases => {
+      discrim = read_4_bytes(ptr)
+      case = $cases[discrim]
+      ptr = align_up_to(ptr + 4, align(case))
+      payload = read_from_memory(case, ptr)
+
+      (discrim, payload) as $language_enum
+    }
+
+    shared-handle => lift(direction, type, [read_4_bytes(ptr)])
+
+    push-buffer | pull-buffer => {
+        assert(direction == "export")
+        lift(direction, type, [read_4_bytes(ptr)])
+    }
+```
+
+## Lowering values
+
+Lowering is very similar to lifting, it's basically just the opposite in various
+directions.
+
+```
+def lower_all(direction, types, values):
+    values = values.iter()
+    types.map(|ty| lower(direction, ty, values))
+
+def lower(direction, type, value):
+  match type
+    f32 => [value as wasm_f32]
+    f64 => [value as wasm_f64]
+    s8 => [value as wasm_i32]
+    u8 => [value as wasm_i32]
+    s16 => [value as wasm_i32]
+    u16 => [value as wasm_i32]
+    s32 => [value as wasm_i32]
+    u32 => [value as wasm_i32]
+    s64 => [value as wasm_i64]
+    u64 => [value as wasm_i64]
+    char => [value as wasm_i32]
+
+    # It's the responsibility of the native language to supply a list where the
+    # in-memory representation of the list matches the canonical ABI.
+    list $ty => {
+      [value.ptr as wasm_i32, value.len as wasm_i32]
+    }
+
+    record $fields => {
+        if is_bitflags($fields):
+          if $fields.len() > 32:
+              [value.bits as wasm_i64]
+          else:
+              [value.bits as wasm_i32]
+        else:
+	  wasm_vals = []
+	  for field in $fields:
+	    wasm_vals.extend(lower(direction, field, value[field.name]))
+          return wasm_vals
+    }
+
+    variant $cases => {
+      match value {
+        the_case(payload)=> {
+          # Recursively lower the optional payload
+          payload_vals = lower(direction, $cases[the_case], payload)
```

Comment:
Since it's optional, I think maybe this goes in an if (since `lower` takes a non-optional type)?